staging/vhost-device-console/src/vhu_console.rs

```diff
@@ -0,0 +1,1412 @@
+        trace!("id 0x{:x}", control_msg.id.to_native());
+        trace!("event 0x{:x}", control_msg.event.to_native());
+        trace!("value 0x{:x}", control_msg.value.to_native());
+    }
+
+    fn process_rx_requests(
+        &mut self,
+        requests: Vec<ConsoleDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        trace!("process_rx_requests");
+
+        if requests.is_empty() {
+            trace!("requests.is_empty");
+            return Ok(true);
+        }
+
+        let desc_chain = &requests[0];
+        let descriptors: Vec<_> = desc_chain.clone().collect();
+
+        if descriptors.len() != 1 {
+            return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+        }
+
+        let desc_request = descriptors[0];
+        if !desc_request.is_write_only() {
+            return Err(Error::UnexpectedReadableDescriptor(0));
+        }
+
+        let response: String = match self.rx_data_fifo.remove() {
+            Ok(item) => item,
+            _ => {
+                trace!("No rx elements");
+                return Ok(false);
+            }
+        };
+
+        desc_chain
+            .memory()
+            .write_slice(response.as_bytes(), desc_request.addr())
+            .map_err(|_| Error::DescriptorWriteFailed)?;
+
+        if vring
+            .add_used(desc_chain.head_index(), response.as_bytes().len() as u32)
+            .is_err()
+        {
+            warn!("Couldn't return used descriptors to the ring");
+        }
+
+        Ok(true)
+    }
+
+    fn process_tx_requests(
+        &mut self,
+        requests: Vec<ConsoleDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<()> {
+        trace!("process_tx_requests");
+
+        if requests.is_empty() {
+            trace!("requests.is_empty");
+            return Ok(());
+        }
+
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+
+            if descriptors.len() != 1 {
+                return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+            }
+
+            let desc_request = descriptors[0];
+            if desc_request.is_write_only() {
+                return Err(Error::UnexpectedWriteOnlyDescriptor(0));
+            }
+
+            let desc_len = desc_request.len();
+            let mut buffer: Vec<u8> = vec![0; desc_len as usize];
+
+            desc_chain
+                .memory()
+                .read_slice(&mut buffer, desc_request.addr())
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            let my_string = String::from_utf8(buffer).unwrap();
+            trace!("{}", my_string);
+            if self.controller.read().unwrap().backend == BackendType::Nested {
+                print!("{}", my_string);
+                io::stdout().flush().unwrap();
+            } else {
+                self.output_buffer = my_string.clone();
```

Comment:
I am probably missing something here but can you explain what this line does? Is it copying the content of each descriptor to `output_buffer`?

Reply:
This was updated to a queue in the latest implementation. Basically I wanted a way to store all the received data from the VM in queue and exit the process function as soon as possible. At a later time,  the backend thread which writes and reads the tcp_socket consumes them asynchronously.